# Day 2: Introduction to Metadata

This notebook introduces the concept of **metadata** using a small teaching dataset.  
It is designed for **classroom presentation**, so each step is short, visual, and easy to explain.

## Learning goals
By the end of this notebook, students should be able to:
- understand what metadata is
- distinguish **data** from **metadata**
- identify **descriptive**, **structural**, and **administrative** metadata
- explain how metadata supports **Findability** and **Reusability**
- connect a simple CSV dataset to its metadata description


In [1]:
# Notebook bootstrap: download required files from GitHub if missing
from pathlib import Path
from urllib.request import urlretrieve

GITHUB_RAW_BASE = "https://raw.githubusercontent.com/MehrdadJalali-AI/data-management-teaching-materials/main"

required_files = [
    ("data/raw/climate_sample.csv", "data/raw/climate_sample.csv"),
]

for local_rel, github_rel in required_files:
    local_path = Path(local_rel)
    if not local_path.exists():
        local_path.parent.mkdir(parents=True, exist_ok=True)
        url = f"{GITHUB_RAW_BASE}/{github_rel}"
        urlretrieve(url, local_path)
        print(f"Downloaded: {local_path}")

print("Bootstrap complete.")

Bootstrap complete.


In [2]:
# Step 0 — Load the dataset
# Keep the original repository file path.
import pandas as pd
from pathlib import Path
from IPython.display import display, Markdown

repo_root = Path().resolve()
csv_path = repo_root / "data" / "raw" / "climate_sample.csv"

print("Expected CSV path:", csv_path)
print("File exists:", csv_path.exists())

df = pd.read_csv(csv_path)
display(df)


Expected CSV path: /Users/mehrdadjalali/Downloads/data/raw/climate_sample.csv
File exists: True


,country,year,avg_temperature,rainfall_mm,co2_emissions
0,Alandia,2021,14.2,820,3.10
1,Alandia,2022,14.6,795,3.25
2,Alandia,2023,15.0,760,3.40
3,Borealia,2021,9.8,1020,2.55
4,Borealia,2022,10.3,980,2.60
5,Borealia,2023,10.9,940,2.70


## Step 1 — First look at the dataset

This is the **data itself**.  
It contains values such as country, year, temperature, rainfall, and CO₂ emissions.

### In class
Ask students:
- What does each row seem to represent?
- What does each column mean?
- Could another person understand this dataset immediately?


In [3]:
# Step 2 — Quick inspection of the dataset
print("Shape of dataset:", df.shape)
print("\nColumns:")
for col in df.columns:
    print("-", col)

print("\nData types:")
display(df.dtypes.to_frame("dtype"))

print("\nBasic summary:")
display(df.describe(include="all").transpose())


Shape of dataset: (6, 5)

Columns:
- country
- year
- avg_temperature
- rainfall_mm
- co2_emissions

Data types:


,dtype
country,object
year,int64
avg_temperature,float64
rainfall_mm,int64
co2_emissions,float64



Basic summary:


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
country,6,2,Alandia,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN
year,6.0,NaN,NaN,NaN,2022.0,0.894427,2021.0,2021.25,2022.0,2022.75,2023.0
avg_temperature,6.0,NaN,NaN,NaN,12.466667,2.376272,9.8,10.45,12.55,14.5,15.0
rainfall_mm,6.0,NaN,NaN,NaN,885.833333,107.90814,760.0,801.25,880.0,970.0,1020.0
co2_emissions,6.0,NaN,NaN,NaN,2.933333,0.362859,2.55,2.625,2.9,3.2125,3.4


## Step 2 — Why raw data alone is not enough

Even if the dataset is small and readable, several questions remain:

- Is this a real or fictional dataset?
- What exactly does `avg_temperature` mean?
- What unit is used for `co2_emissions`?
- Who created the dataset?
- Can others reuse it legally?
- What does one row represent?

These answers are usually not stored in the data values themselves.  
They are provided by **metadata**.


## Step 3 — Define three metadata types

For teaching purposes, we will describe the dataset using three simple metadata layers:

### 1. Descriptive metadata
Helps people **find** and understand what the dataset is about.

### 2. Structural metadata
Explains how the dataset is **organized**.

### 3. Administrative metadata
Describes management, ownership, access, versioning, and reuse conditions.


In [4]:
# Step 3A — Descriptive metadata
descriptive_metadata = {
    "title": "2025 Global Climate Data",
    "description": "Fictional annual climate summaries for two countries from 2021 to 2023.",
    "keywords": ["climate", "temperature", "rainfall", "CO2 emissions"],
    "creator": "Teaching repository example",
    "subject": "Climate and environmental indicators"
}

display(pd.DataFrame(
    {"Field": list(descriptive_metadata.keys()),
     "Value": [str(v) for v in descriptive_metadata.values()]}
))


,Field,Value
0,title,2025 Global Climate Data
1,description,Fictional annual climate summaries for two cou...
2,keywords,"['climate', 'temperature', 'rainfall', 'CO2 em..."
3,creator,Teaching repository example
4,subject,Climate and environmental indicators


In [5]:
# Step 3B — Structural metadata
structural_metadata = {
    "rows": "Each row represents one country-year observation.",
    "columns": {
        "country": "Country name",
        "year": "Year of observation",
        "avg_temperature": "Average temperature in degrees Celsius",
        "rainfall_mm": "Annual rainfall in millimeters",
        "co2_emissions": "CO2 emissions in fictional units",
    },
}

structural_rows = [{"Column": k, "Meaning": v} for k, v in structural_metadata["columns"].items()]
display(pd.DataFrame(structural_rows))


,Column,Meaning
0,country,Country name
1,year,Year of observation
2,avg_temperature,Average temperature in degrees Celsius
3,rainfall_mm,Annual rainfall in millimeters
4,co2_emissions,CO2 emissions in fictional units


In [6]:
# Step 3C — Administrative metadata
administrative_metadata = {
    "license": "CC BY 4.0 (example)",
    "version": "1.0 (teaching dataset)",
    "contact": "research-data@example.org (example)",
    "access": "Open for teaching purposes",
    "provenance_note": "Created as a fictional classroom dataset"
}

display(pd.DataFrame(
    {"Field": list(administrative_metadata.keys()),
     "Value": [str(v) for v in administrative_metadata.values()]}
))


,Field,Value
0,license,CC BY 4.0 (example)
1,version,1.0 (teaching dataset)
2,contact,research-data@example.org (example)
3,access,Open for teaching purposes
4,provenance_note,Created as a fictional classroom dataset


## Step 4 — Data vs. Metadata

A simple way to explain it in class:

- **Data** = the actual values  
  Example: `Germany, 2023, 14.2, 620, 11.3`

- **Metadata** = information about the data  
  Example:  
  - title of the dataset  
  - meaning of each column  
  - license  
  - creator  
  - version

### Teaching analogy
A dataset is like a **book**.  
Metadata is like the **catalog entry** and **front matter** that tell us:
- title
- author
- subject
- edition
- access conditions


In [7]:
# Step 5 — Combine metadata into one classroom view
metadata_overview = [
    {"Type": "Descriptive", "Purpose": "Discovery and understanding", "Example": "Title, description, keywords"},
    {"Type": "Structural", "Purpose": "Organization and interpretation", "Example": "Rows, columns, units"},
    {"Type": "Administrative", "Purpose": "Management and reuse", "Example": "License, version, contact"},
]

display(pd.DataFrame(metadata_overview))


,Type,Purpose,Example
0,Descriptive,Discovery and understanding,"Title, description, keywords"
1,Structural,Organization and interpretation,"Rows, columns, units"
2,Administrative,Management and reuse,"License, version, contact"


## Step 6 — Why metadata matters for FAIR

Metadata is essential for FAIR data practices.

### Findable
A dataset becomes easier to find when it has:
- a clear title
- useful keywords
- a good description
- ideally a persistent identifier

### Reusable
A dataset becomes easier to reuse when it has:
- clear column meanings
- units and structure
- license information
- version and provenance details


In [8]:
# Step 6 — Connect our example metadata to FAIR
fair_mapping = [
    {"FAIR principle": "Findable", "Helpful metadata fields": "title, description, keywords, creator"},
    {"FAIR principle": "Reusable", "Helpful metadata fields": "column meanings, license, version, provenance"},
]

display(pd.DataFrame(fair_mapping))


,FAIR principle,Helpful metadata fields
0,Findable,"title, description, keywords, creator"
1,Reusable,"column meanings, license, version, provenance"


## Step 7 — Small classroom discussion

Discuss with students:

1. Which metadata fields help most with **finding** the dataset?
2. Which metadata fields help most with **understanding** the dataset?
3. Which metadata fields help most with **reusing** the dataset?
4. What important metadata is still missing?


In [9]:
# Step 7 — Identify possible weaknesses in the current metadata
weaknesses = [
    "No DOI or other persistent identifier is provided.",
    "No publication date is explicitly stated.",
    "The geographic scope is not fully described.",
    "The metadata does not define the exact unit for CO2 emissions in a standard way.",
    "No detailed provenance or collection method is given."
]

display(pd.DataFrame({"Potential metadata weakness": weaknesses}))


,Potential metadata weakness
0,No DOI or other persistent identifier is provi...
1,No publication date is explicitly stated.
2,The geographic scope is not fully described.
3,The metadata does not define the exact unit fo...
4,No detailed provenance or collection method is...


## Step 8 — Key takeaway

This notebook showed that even a very small CSV file benefits from metadata.

### Final message for students
A dataset without metadata may still be readable, but it is much harder to:
- discover
- interpret
- trust
- reuse

In the next notebooks, we will move from these **conceptual metadata examples** to actual metadata standards:
- **Dublin Core**
- **DataCite**
- **schema.org**
